File: 337. House Robber III-1.cpp
class Solution {
public:
  int rob(TreeNode *root) {
    pair<int, int> res = helper(root);

    return max(res.first, res.second);
  }
  pair<int, int> helper(TreeNode *root) {
    if (!root)
      return {0, 0};

    pair<int, int> l = helper(root->left);
    pair<int, int> r = helper(root->right);

    int t = l.second + r.second + root->val;
    int nt = max(l.first, l.second) + max(r.first, r.second);

    return {t, nt};
  }
};

File: 337. House Robber III-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            if root is None:
                return 0, 0
            l_rob, l_not_rob = dfs(root.left)
            r_rob, r_not_rob = dfs(root.right)
            rob = l_not_rob + r_not_rob + root.val
            not_rob = max(l_rob, l_not_rob) + max(r_rob, r_not_rob)
            return rob, not_rob

        return max(dfs(root))
